src/Illuminate/Database/Eloquent/Model.php

```diff
@@ -2808,6 +2808,10 @@ protected function castAttribute($key, $value)
                 return json_decode($value, true);
             case 'collection':
                 return new BaseCollection(json_decode($value, true));
+            case 'date':
+                return Carbon::createFromFormat('Y-m-d', $value);
+            case 'datetime':
+                return Carbon::createFromFormat('Y-m-d H:i:s', $value);
```

Comment:
And used in Eloquent via https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Eloquent/Model.php#L2946